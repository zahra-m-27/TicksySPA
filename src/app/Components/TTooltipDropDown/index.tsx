import React, {useEffect, useRef, useState} from 'react';
import Assets from '../../Assets';
import styles from './styles.module.scss';
import ClassNames from '../../Utilities/ClassNames';

interface Item<T> {
  value: T;
  label: string;
}

interface Props<T> {
  items: Item<T>[];
  className?: string;
  selectedItem?: T;
  children: React.ReactNode;
  onSelect?: (value: T) => void;
}

export default function TTooltipDropDown<T>({
  items,
  children,
  onSelect,
  className,
  selectedItem,
}: Props<T>) {
  const [Open, setOpen] = useState<boolean>(false);
  const container = useRef<HTMLDivElement>(null);
  const content = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const onClickContainer = () => setOpen(false);
    const onClickContent = (e: any) => {
      e.stopPropagation();
      setOpen(!Open);
    };
    container.current?.parentElement?.addEventListener(
      'click',
      onClickContainer,
      false
    );
    content.current?.addEventListener('click', onClickContent, false);
    return () => {
      container.current?.parentElement?.removeEventListener(
        'click',
        onClickContainer
      );
      content.current?.removeEventListener('click', onClickContent);
    };
  }, [Open]);

  const container_style = ClassNames(styles.container, className);

  return (
    <div ref={container} data-open={Open} className={container_style}>
      <div ref={content}>{children}</div>

      <div className={styles.items}>
        {items.map((item) => (
          <div
            key={`${item.value}`}
            className={styles.item}
            onClick={() => {
              setOpen(false);
              onSelect && onSelect(item.value);
            }}>
            <span>{item.label}</span>
            {selectedItem === item.value && (
              <div className={styles.checkmark}></div>
            )}
          </div>
        ))}
      </div>
    </div>
  );
}
