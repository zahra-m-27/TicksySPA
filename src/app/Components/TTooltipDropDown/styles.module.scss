.container {
  z-index: 1;
  cursor: pointer;
  position: relative;

  .arrow_down {
  }

  &[data-open='false'] {
    .items {
      display: none;
    }
  }

  .selected_item_label {
    color: #606f7b;
    cursor: pointer;
  }

  .items {
    padding: 10px;
    cursor: pointer;
    transition: 0.6s;
    width: max-content;
    position: absolute;
    border-radius: 8px;
    background-color: #eeeff1;
    box-shadow: 0 3px 6px 0 rgb(0 0 0 / 16%);

    .item {
      color: #000;
      margin: 0.5px;
      display: flex;
      text-align: end;
      padding: 8px 10px;
      align-items: center;
      transition: all 0.3s;
      flex-direction: row-reverse;

      &:hover {
        overflow: hidden;
        position: relative;
        border-radius: 5px;
        background-color: #d5e1e6;
      }

      span {
        flex: 1;
        margin-left: 10px;
      }

      .checkmark {
        width: 15px;
        height: 15px;
        border-radius: 50%;
        border: 3px solid #00b9be;
      }
    }
  }
}
