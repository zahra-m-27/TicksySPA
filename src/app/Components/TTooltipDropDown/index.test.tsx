import React from 'react';
import {render} from '@testing-library/react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import user from '@testing-library/user-event';
import TTooltipDropDown from './index';

const onSelect = jest.fn();

const sample = (
  <TTooltipDropDown
    onSelect={onSelect}
    items={[
      {
        value: 'val1',
        label: 'تست1',
      },
      {
        value: 'val2',
        label: 'تست2',
      },
      {
        value: 'val3',
        label: 'تست3',
      },
    ]}>
    <div></div>
  </TTooltipDropDown>
);

test('render in ReactDOM', () => {
  const div = document.createElement('div');
  ReactDOM.render(sample, div);
});

test('render items', () => {
  const {getByText} = render(sample);
  getByText(/تست1/i);
  getByText(/تست2/i);
  getByText(/تست3/i);
});

test('onSelect drop down', () => {
  const {getByText} = render(sample);
  const item1 = getByText(/تست1/i);
  user.click(item1);
  expect(onSelect).toBeCalledTimes(1);
  expect(onSelect).toBeCalledWith('val1');
});

it('snapshot', () => {
  const snapshot = renderer.create(sample).toJSON();
  expect(snapshot).toMatchSnapshot();
});
