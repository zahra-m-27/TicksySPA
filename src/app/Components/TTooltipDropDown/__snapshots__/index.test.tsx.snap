// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="container"
  data-open={false}
>
  <div>
    <div />
  </div>
  <div
    className="items"
  >
    <div
      className="item"
      onClick={[Function]}
    >
      <span>
        تست1
      </span>
    </div>
    <div
      className="item"
      onClick={[Function]}
    >
      <span>
        تست2
      </span>
    </div>
    <div
      className="item"
      onClick={[Function]}
    >
      <span>
        تست3
      </span>
    </div>
  </div>
</div>
`;
